File: pickers/src/main/java/com/rands/lightbulb/pickers/dialog/base/BaseDialogPickerView.java
package com.rands.lightbulb.pickers.dialog.base;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.Log;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.rands.lightbulb.dialogs.base.BasePickerDialogFragment;
import com.rands.lightbulb.pickers.R;

import java.util.ArrayList;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import static com.rands.lightbulb.dialogs.base.BaseDialogFragment.DialogAnimationTypes;

@SuppressWarnings({"unused", "UnusedReturnValue", "rawtypes"})
public abstract class BaseDialogPickerView extends LinearLayout {
    protected DialogPickerTriggerLayout triggerView;
    protected String DIALOG_TAG;
    protected String pickerHintText;
    protected String errorText;
    protected String pickerDialogPositiveButtonText;
    protected String pickerDialogNegativeButtonText;
    protected String pickerRequiredText;
    protected boolean errorEnabled;
    protected boolean validationEnabled;
    protected boolean required;
    protected boolean showSelectedTextValue;
    protected boolean pickerDialogCancelable;
    protected Drawable pickerIcon;
    protected DialogAnimationTypes pickerDialogAnimationType;
    protected FragmentManager manager;
    protected BasePickerDialogFragment pickerDialog;
    private final ArrayList<TriggerAttachedCallback> triggerAttachedCallback = new ArrayList<>();

    public BaseDialogPickerView(Context context) {
        this(context, null);
    }

    public BaseDialogPickerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setSaveEnabled(true);
        readBaseAttributes(context, attrs);
        readAttributes(context, attrs);
        if (!isInEditMode())
            manager = ((FragmentActivity) context).getSupportFragmentManager();
        setOrientation(VERTICAL);
    }

    public abstract boolean validate();

    protected abstract BasePickerDialogFragment initializeDialog();

    protected abstract BasePickerDialogFragment getDialog();

    protected abstract String getViewText();

    protected abstract void readAttributes(Context context, AttributeSet attrs);

    @Override
    public void addView(View child, int index, ViewGroup.LayoutParams params) {
        if (!(child instanceof DialogPickerTriggerLayout)) {
            Log.w(BaseDialogPickerView.class.getName(), "Picker view child is ignored. All child views must implement com.rands.lightbulb.pickers.dialog.base.TriggerView");
            return;
        }
        if (getChildCount() > 0) {
            Log.w(BaseDialogPickerView.class.getName(), "Picker can have only one trigger view.");
            return;
        }
        super.addView(child, index, params);
        triggerView = (DialogPickerTriggerLayout) child;
        triggerView.attachTo(this);
        for (TriggerAttachedCallback callback : triggerAttachedCallback)
            callback.onAttached(this.triggerView, this);
    }

    private void readBaseAttributes(Context context, AttributeSet attrs) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.PickerView, 0, 0);
        try {
            pickerHintText = a.getString(R.styleable.PickerView_PV_HintText);
            errorText = a.getString(R.styleable.PickerView_PV_ErrorText);
            pickerRequiredText = a.getString(R.styleable.PickerView_PV_RequiredText);
            pickerDialogPositiveButtonText = a.getString(R.styleable.PickerView_PV_DialogButtonPosText);
            pickerDialogNegativeButtonText = a.getString(R.styleable.PickerView_PV_DialogButtonNegText);
            DIALOG_TAG = a.getString(R.styleable.PickerView_PV_DialogTag);
            if (pickerHintText == null || pickerHintText.equals(""))
                pickerHintText = "";
            if (errorText == null || errorText.equals(""))
                errorText = "Error";
            if (pickerDialogPositiveButtonText == null || pickerDialogPositiveButtonText.equals(""))
                pickerDialogPositiveButtonText = "OK";
            if (pickerDialogNegativeButtonText == null || pickerDialogNegativeButtonText.equals(""))
                pickerDialogNegativeButtonText = "CANCEL";
            if (pickerRequiredText == null || pickerRequiredText.equals(""))
                pickerRequiredText = "Field is required";
            if (DIALOG_TAG == null || DIALOG_TAG.equals(""))
                DIALOG_TAG = "DEFAULT_PICKER_DIALOG_TAG";
            pickerDialogAnimationType = DialogAnimationTypes.valueOf(a.getInt(R.styleable.PickerView_PV_DialogAnimation, 1));
            errorEnabled = a.getBoolean(R.styleable.PickerView_PV_ErrorEnabled, false);
            required = a.getBoolean(R.styleable.PickerView_PV_Required, false);
            validationEnabled = a.getBoolean(R.styleable.PickerView_PV_ValidationEnabled, false);
            pickerDialogCancelable = a.getBoolean(R.styleable.PickerView_PV_DialogCancelable, true);
            showSelectedTextValue = a.getBoolean(R.styleable.PickerView_PV_ShowSelectedText, true);
        } finally {
            a.recycle();
        }
    }

    protected final void updateTextAndValidate() {
        updateText();
        validate();
    }

    protected final void updateText() {
        if (!showSelectedTextValue)
            return;
        String newTextValue = this.getViewText();
        triggerView.setTriggerText(newTextValue);
    }

    public final void showPickerDialog() {
        if (pickerDialog == null)
            pickerDialog = initializeDialog();
        pickerDialog.show();
    }

    public void addOnTriggerAttachedCallback(TriggerAttachedCallback callback) {
        triggerAttachedCallback.add(callback);
    }

    public final void setTriggerView(View triggerView) {
        if (!(triggerView instanceof DialogPickerTriggerLayout)) {
            Log.w(BaseDialogPickerView.class.getName(), "Trigger view must implement com.rands.lightbulb.pickers.dialog.base.TriggerView");
            return;
        }
        removeAllViews();
        addView(triggerView);
        this.triggerView = (DialogPickerTriggerLayout) triggerView;
        this.triggerView.attachTo(this);
        for (TriggerAttachedCallback callback : triggerAttachedCallback)
            callback.onAttached(this.triggerView, this);
    }

    public final void setValidationEnabled(boolean validationEnabled) {
        this.validationEnabled = validationEnabled;
    }

    public final void setDialogTag(String dialogTag) {
        if (pickerDialog == null)
            this.DIALOG_TAG = dialogTag;
    }

    public final void setDialogAnimation(DialogAnimationTypes animation) {
        this.pickerDialogAnimationType = animation;
    }

    public final void setRequired(boolean required) {
        this.required = required;
        if (required)
            validate();
    }

    public final void setErrorText(String error) {
        errorText = error;
        triggerView.setTriggerErrorText(errorText);
    }

    public final void setHintText(String hintText) {
        if (!showSelectedTextValue)
            return;
        pickerHintText = hintText;
        triggerView.setTriggerHintText(pickerHintText);
    }

    public final void setPickerIcon(Drawable icon, Integer color) {
        pickerIcon = icon;
        triggerView.setTriggerIcon(icon, color);
    }

    public final void setPickerIcon(Drawable icon) {
        setPickerIcon(icon, null);
    }

    public final void setShowSelectedTextValue(boolean showSelectedTextValue) {
        this.showSelectedTextValue = showSelectedTextValue;
        invalidate();
    }

    protected final void setErrorEnabled(boolean errorEnabled) {
        if (this.errorEnabled != errorEnabled) {
            this.errorEnabled = errorEnabled;
            triggerView.setTriggerErrorEnabled(this.errorEnabled);
        }
    }

    protected final boolean isPickerDialogShowing() {
        return pickerDialog != null && pickerDialog.isDialogShown();
    }

    public final boolean isShowSelectedTextValue() {
        return showSelectedTextValue;
    }

    public final boolean isValidationEnabled() {
        return validationEnabled;
    }

    public final boolean isErrorEnabled() {
        return this.errorEnabled;
    }

    public final String getPickerHintText() {
        return pickerHintText;
    }

    public final String getErrorText() {
        return errorText;
    }

    public final String getText() {
        return triggerView.getTriggerText();
    }

    public final DialogPickerTriggerLayout getTriggerView() {
        return triggerView;
    }

    @Override
    protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
        dispatchFreezeSelfOnly(container);
    }

    @Override
    protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
        dispatchThawSelfOnly(container);
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState myState = new SavedState(superState);
        myState.isDialogShown = isPickerDialogShowing();
        myState.pickerHintText = pickerHintText;
        myState.pickerErrorText = errorText;
        myState.pickerRequiredText = pickerRequiredText;
        myState.pickerIsRequired = required;
        myState.pickerIsErrorEnabled = errorEnabled;
        myState.pickerIsValidationEnabled = validationEnabled;
        myState.pickerShowSelectedTextValue = showSelectedTextValue;
        myState.pickerDialogAnimationType = pickerDialogAnimationType.getValue();
        myState.pickerDialogPositiveButtonText = pickerDialogPositiveButtonText;
        myState.pickerDialogNegativeButtonText = pickerDialogNegativeButtonText;
        myState.pickerDialogCancelable = pickerDialogCancelable;
        return myState;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        SavedState savedState = (SavedState) state;
        super.onRestoreInstanceState(savedState.getSuperState());
        boolean isPickerDialogShowing = savedState.isDialogShown;
        pickerHintText = savedState.pickerHintText;
        errorText = savedState.pickerErrorText;
        pickerRequiredText = savedState.pickerRequiredText;
        required = savedState.pickerIsRequired;
        errorEnabled = savedState.pickerIsErrorEnabled;
        validationEnabled = savedState.pickerIsValidationEnabled;
        showSelectedTextValue = savedState.pickerShowSelectedTextValue;
        pickerDialogAnimationType = DialogAnimationTypes.valueOf(savedState.pickerDialogAnimationType);
        pickerDialogPositiveButtonText = savedState.pickerDialogPositiveButtonText;
        pickerDialogNegativeButtonText = savedState.pickerDialogNegativeButtonText;
        pickerDialogCancelable = savedState.pickerDialogCancelable;
        updateTextAndValidate();
        if (isPickerDialogShowing)
            showPickerDialog();
    }

    private static class SavedState extends BaseSavedState {
        private boolean isDialogShown;
        private String pickerHintText;
        private String pickerErrorText;
        private String pickerRequiredText;
        private boolean pickerIsRequired;
        private boolean pickerIsErrorEnabled;
        private boolean pickerIsValidationEnabled;
        private boolean pickerShowSelectedTextValue;
        private int pickerDialogAnimationType;
        private int pickerDialogAnimationDuration;
        private String pickerDialogPositiveButtonText;
        private String pickerDialogNegativeButtonText;
        private boolean pickerDialogCancelable;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            isDialogShown = in.readByte() != 0;
            pickerHintText = in.readString();
            pickerErrorText = in.readString();
            pickerRequiredText = in.readString();
            pickerIsRequired = in.readByte() != 0;
            pickerIsErrorEnabled = in.readByte() != 0;
            pickerIsValidationEnabled = in.readByte() != 0;
            pickerShowSelectedTextValue = in.readByte() != 0;
            pickerDialogAnimationType = in.readInt();
            pickerDialogAnimationDuration = in.readInt();
            pickerDialogPositiveButtonText = in.readString();
            pickerDialogNegativeButtonText = in.readString();
            pickerDialogCancelable = in.readByte() != 0;
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeByte((byte) (isDialogShown ? 1 : 0));
            out.writeString(pickerHintText);
            out.writeString(pickerErrorText);
            out.writeString(pickerRequiredText);
            out.writeByte((byte) (pickerIsRequired ? 1 : 0));
            out.writeByte((byte) (pickerIsErrorEnabled ? 1 : 0));
            out.writeByte((byte) (pickerIsValidationEnabled ? 1 : 0));
            out.writeByte((byte) (pickerShowSelectedTextValue ? 1 : 0));
            out.writeInt(pickerDialogAnimationType);
            out.writeInt(pickerDialogAnimationDuration);
            out.writeString(pickerDialogPositiveButtonText);
            out.writeString(pickerDialogNegativeButtonText);
            out.writeByte((byte) (pickerDialogCancelable ? 1 : 0));
        }

        public static final Creator<SavedState> CREATOR
                = new Creator<SavedState>() {
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    public interface TriggerAttachedCallback {
        void onAttached(DialogPickerTriggerLayout triggerView, BaseDialogPickerView pickerView);
    }
}

File: lightbulb-pickers/src/main/java/com/github/rooneyandshadows/lightbulb/pickers/dialog/base/BaseDialogPickerView.kt
package com.github.rooneyandshadows.lightbulb.pickers.dialog.base

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.util.AttributeSet
import android.util.Log
import android.util.SparseArray
import android.view.ContextThemeWrapper
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.fragment.app.FragmentManager
import com.github.rooneyandshadows.lightbulb.commons.utils.ParcelUtils
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.Buttons.Companion.cancelSelectionButton
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.Buttons.Companion.confirmSelectionButton
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogAnimationTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogAnimationTypes.NO_ANIMATION
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogButton
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes.NORMAL
import com.github.rooneyandshadows.lightbulb.pickers.R
import com.github.rooneyandshadows.lightbulb.pickers.dialog.base.BaseDialogPickerView.ButtonTypes.NEGATIVE
import com.github.rooneyandshadows.lightbulb.pickers.dialog.base.BaseDialogPickerView.ButtonTypes.POSITIVE
import com.github.rooneyandshadows.lightbulb.pickers.dialog.trigger.base.DialogTriggerView

@Suppress("unused", "MemberVisibilityCanBePrivate")
@JvmSuppressWildcards
abstract class BaseDialogPickerView<SelectionType> @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : LinearLayoutCompat(context, attrs, defStyleAttr) {
    private lateinit var fragmentManager: FragmentManager
    private lateinit var dialogTag: String
    private val validationCallbacks: MutableList<ValidationCheck<SelectionType>> = mutableListOf()
    private val selectionChangedListeners: MutableList<SelectionChangedListener<SelectionType>> = mutableListOf()
    private val triggerAttachedCallback: MutableList<TriggerAttachedCallback<SelectionType>> = mutableListOf()
    private val onDialogReadyListeners: MutableList<DialogReadyListener<SelectionType>> = mutableListOf()
    private val dialogInitializer = lazy {
        return@lazy getDialogBuilder(fragmentManager, dialogTag).buildDialog().apply dialog@{
            onDialogReadyListeners.apply {
                forEach { it.execute(this@dialog) }
                clear()
            }
        }
    }
    private val pickerDialog: BasePickerDialogFragment<SelectionType> by dialogInitializer
    private var positiveButtonText: String = ""
    private var negativeButtonText: String = ""
    protected var dataBindingListener: SelectionChangedListener<SelectionType>? = null
    protected var triggerView: DialogTriggerView? = null
        private set
    var showSelectedTextValue = false
        protected set
    var requiredText: String? = null
    var isValidationEnabled = false
        private set
    var isRequired = false
        private set
    val isDialogShown: Boolean
        get() = pickerDialog.isDialogShown
    val text: String
        get() = triggerView?.text ?: ""
    val hasSelection: Boolean
        get() = pickerDialog.hasSelection()
    protected open val dialog: BasePickerDialogFragment<SelectionType>
        get() = pickerDialog
    open var selection: SelectionType?
        set(value) = pickerDialog.setSelection(value)
        get() = pickerDialog.getSelection()
    protected abstract val viewText: String

    protected abstract fun getDialogBuilder(
        fragmentManager: FragmentManager,
        dialogTag: String,
    ): BaseDialogBuilder<out BasePickerDialogFragment<SelectionType>>

    init {
        isSaveEnabled = true
        orientation = VERTICAL
        if (!isInEditMode) fragmentManager = getFragmentManager(context)!!
        readBaseAttributes(context, attrs)
        whenDialogReady { dialog ->
            dialog.apply {
                if (negativeButtonText.isNotBlank()) {
                    addButton(generateButtonConfig(NEGATIVE))
                }
                if (positiveButtonText.isNotBlank()) {
                    addButton(generateButtonConfig(POSITIVE))
                }
                addOnCancelListener { updateTextAndValidate() }
                addOnSelectionChangedListener(object : BasePickerDialogFragment.SelectionChangedListener<SelectionType> {
                    override fun onSelectionChanged(
                        dialog: BasePickerDialogFragment<SelectionType>,
                        newValue: SelectionType?,
                        oldValue: SelectionType?,
                    ) {
                        updateTextAndValidate()
                        dispatchSelectionChangedEvents(newValue, oldValue)
                    }
                })
            }
        }
    }

    fun validate(): Boolean {
        var isValid = true
        if (isValidationEnabled) {
            if (isRequired && !hasSelection) {
                setErrorEnabled(true)
                setErrorText(requiredText)
                return false
            }
            validationCallbacks.forEach {
                isValid = isValid and it.validate(selection)
                if (!isValid) return@forEach
            }
        }
        if (!isValid) setErrorEnabled(true)
        else {
            setErrorEnabled(false)
            setErrorText(null)
        }
        return isValid
    }

    private fun setErrorEnabled(errorEnabled: Boolean) {
        triggerView?.setErrorEnabled(errorEnabled)
    }

    private fun getErrorEnabled(): Boolean {
        return triggerView?.errorEnabled ?: false
    }

    @Override
    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        pickerDialog.apply {}//init dialog
    }

    @Override
    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        triggerView?.isEnabled = enabled
    }

    @Override
    override fun addView(child: View, index: Int, params: ViewGroup.LayoutParams) {
        addViewInternally(child, index, params)
    }

    fun addSelectionChangedListener(listener: SelectionChangedListener<SelectionType>) {
        selectionChangedListeners.add(listener)
    }

    fun removeSelectionChangedListener(listener: SelectionChangedListener<SelectionType>) {
        selectionChangedListeners.add(listener)
    }

    fun removeAllSelectionChangedListeners() {
        selectionChangedListeners.clear()
    }

    fun addValidationCheck(validationCheck: ValidationCheck<SelectionType>) {
        validationCallbacks.add(validationCheck)
    }

    fun removeValidationCheck(validationCheck: ValidationCheck<SelectionType>) {
        validationCallbacks.remove(validationCheck)
    }

    fun removeAllValidationChecks() {
        validationCallbacks.clear()
    }

    fun addOnTriggerAttachedListener(callback: TriggerAttachedCallback<SelectionType>) {
        triggerAttachedCallback.add(callback)
    }

    fun showPickerDialog() {
        pickerDialog.show()
    }


    fun setDialogTag(dialogTag: String) {
        pickerDialog.setDialogTag(dialogTag)
    }

    fun setDialogTitle(dialogTitle: String?) {
        pickerDialog.setDialogTitle(dialogTitle)
    }

    fun setDialogMessage(dialogMessage: String?) {
        pickerDialog.setDialogMessage(dialogMessage)
    }

    fun setDialogPositiveButtonText(buttonText: String?) {
        val addButton: Boolean = !buttonText.isNullOrBlank()
        positiveButtonText = buttonText ?: ""
        val button = generateButtonConfig(POSITIVE)
        if (addButton) pickerDialog.addButton(button)
        else pickerDialog.removeButton(button)
    }

    fun setDialogNegativeButtonText(buttonText: String?) {
        val addButton: Boolean = !buttonText.isNullOrBlank()
        negativeButtonText = buttonText ?: ""
        val button = generateButtonConfig(NEGATIVE)
        if (addButton) pickerDialog.addButton(button)
        else pickerDialog.removeButton(button)
    }

    fun setDialogCancelable(isCancelable: Boolean) {
        pickerDialog.isCancelable = isCancelable
    }

    fun setDialogType(dialogType: DialogTypes) {
        pickerDialog.dialogType = dialogType
    }

    fun setDialogAnimationType(dialogAnimationType: DialogAnimationTypes) {
        pickerDialog.dialogAnimationType = dialogAnimationType
    }

    fun setErrorText(errorText: String?) {
        triggerView?.apply {
            post {
                setErrorText(errorText)
            }
        }
    }

    fun getDialogTag(): String? {
        return pickerDialog.tag
    }

    fun getDialogTitle(): String? {
        return pickerDialog.dialogTitle
    }

    fun getDialogMessage(): String? {
        return pickerDialog.dialogMessage
    }

    fun isDialogCancelable(): Boolean {
        return pickerDialog.isCancelable
    }

    fun getDialogType(): DialogTypes {
        return pickerDialog.dialogType
    }

    fun getDialogAnimationType(): DialogAnimationTypes {
        return pickerDialog.dialogAnimationType
    }

    fun isErrorEnabled(): Boolean {
        return triggerView?.errorEnabled ?: false
    }

    fun getErrorText(): String? {
        return triggerView?.errorText
    }

    @JvmOverloads
    fun setRequired(required: Boolean, validateOnCall: Boolean = true) {
        this.isRequired = required
        if (validateOnCall) validate()
    }

    fun setValidationEnabled(newValue: Boolean) {
        isValidationEnabled = newValue
        validate()
    }

    @JvmOverloads
    fun setPickerIcon(icon: Drawable?, color: Int = -1) {
        triggerView?.setIcon(icon, color)
    }

    protected fun whenDialogReady(action: DialogReadyListener<SelectionType>) {
        if (dialogInitializer.isInitialized()) {
            action.execute(pickerDialog)
            return
        }
        onDialogReadyListeners.add(action)
    }

    protected fun updateTextAndValidate() {
        updateText()
        validate()
    }

    protected fun setShowSelectionText(showSelectedTextValue: Boolean) {
        this.showSelectedTextValue = showSelectedTextValue
        invalidate()
    }

    private fun getFragmentManager(context: Context?): FragmentManager? {
        return when (context) {
            is AppCompatActivity -> context.supportFragmentManager
            is ContextThemeWrapper -> getFragmentManager(context.baseContext)
            else -> null
        }
    }

    private fun updateText() {
        if (!showSelectedTextValue) return
        val newTextValue = viewText
        triggerView?.setText(newTextValue)
    }

    private fun addViewInternally(child: View, index: Int, params: ViewGroup.LayoutParams) {
        if (child !is DialogTriggerView) {
            Log.w(
                BaseDialogPickerView::class.java.name,
                "Picker view child is ignored. All child views must implement com.github.rooneyandshadows.lightbulb.pickers.dialog.trigger.base.DialogTriggerView"
            )
            return
        }
        removeAllViews()
        triggerView = child
        super.addView(child, index, params)
        triggerView!!.apply {
            isEnabled = this@BaseDialogPickerView.isEnabled
            setText(viewText)
            attachTo(this@BaseDialogPickerView)
            triggerAttachedCallback.forEach {
                it.onAttached(this, this@BaseDialogPickerView)
            }
        }
    }

    private fun generateButtonConfig(buttonType: ButtonTypes): DialogButton {
        return when (buttonType) {
            POSITIVE -> confirmSelectionButton(positiveButtonText) { _, _ -> validate() }
            NEGATIVE -> cancelSelectionButton(negativeButtonText) { _, _ -> validate() }
        }
    }

    private fun readBaseAttributes(context: Context, attrs: AttributeSet?) {
        val attrTypedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.BaseDialogPickerView, 0, 0)
        try {
            attrTypedArray.apply {
                getString(R.styleable.BaseDialogPickerView_pv_dialog_tag).apply {
                    dialogTag = let {
                        val default = ResourceUtils.getPhrase(context, R.string.picker_default_dialog_tag_text)
                        return@let if (it.isNullOrBlank()) default else it
                    }
                }
                getString(R.styleable.BaseDialogPickerView_pv_dialog_title).apply {
                    val default = ""
                    whenDialogReady {
                        it.setDialogTitle(this ?: default)
                    }
                }
                getString(R.styleable.BaseDialogPickerView_pv_dialog_message).apply {
                    val default = ""
                    whenDialogReady {
                        it.setDialogMessage(this ?: default)
                    }
                }
                getString(R.styleable.BaseDialogPickerView_pv_required_text).apply {
                    val default = ResourceUtils.getPhrase(context, R.string.picker_default_required_text)
                    requiredText = this ?: default
                }
                getString(R.styleable.BaseDialogPickerView_pv_dialog_button_positive_text).apply {
                    positiveButtonText = let {
                        val default = ResourceUtils.getPhrase(context, R.string.picker_default_positive_button_text)
                        return@let if (it.isNullOrBlank()) default else it
                    }
                }
                getString(R.styleable.BaseDialogPickerView_pv_dialog_button_negative_text).apply {
                    negativeButtonText = let {
                        val default = ResourceUtils.getPhrase(context, R.string.picker_default_negative_button_text)
                        return@let if (it.isNullOrBlank()) default else it
                    }
                }
                getInt(R.styleable.BaseDialogPickerView_pv_dialog_type, NORMAL.value).apply {
                    whenDialogReady {
                        it.dialogType = DialogTypes.valueOf(this)
                    }
                }
                getInt(R.styleable.BaseDialogPickerView_pv_dialog_animation, NO_ANIMATION.value).apply {
                    whenDialogReady {
                        it.dialogAnimationType = DialogAnimationTypes.valueOf(this)
                    }
                }
                getBoolean(R.styleable.BaseDialogPickerView_pv_dialog_cancelable, true).apply {
                    whenDialogReady {
                        it.isCancelable = this
                    }
                }
                isRequired = getBoolean(R.styleable.BaseDialogPickerView_pv_required, false)
                isValidationEnabled = getBoolean(R.styleable.BaseDialogPickerView_pv_validation_enabled, false)
                showSelectedTextValue = getBoolean(R.styleable.BaseDialogPickerView_pv_show_selected_text, true)
            }
        } finally {
            attrTypedArray.recycle()
        }
    }

    private fun dispatchSelectionChangedEvents(newValue: SelectionType?, oldValue: SelectionType?) {
        selectionChangedListeners.forEach {
            it.execute(newValue, oldValue)
        }
        dataBindingListener?.execute(newValue, oldValue)
    }

    @Override
    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        dispatchFreezeSelfOnly(container)
    }

    @Override
    override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
        dispatchThawSelfOnly(container)
    }

    @Override
    public override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val myState = SavedState(superState)
        myState.apply {
            val view = this@BaseDialogPickerView
            isDialogShown = view.isDialogShown
            requiredText = view.requiredText
            isRequired = view.isRequired
            isValidationEnabled = view.isValidationEnabled
            showSelectedTextValue = view.showSelectedTextValue
            dialogState = pickerDialog.saveDialogState()
            triggerState = triggerView?.onSaveInstanceState()
        }
        return myState
    }

    @Override
    public override fun onRestoreInstanceState(state: Parcelable) {
        val savedState = state as SavedState
        super.onRestoreInstanceState(savedState.superState)
        state.apply {
            val view = this@BaseDialogPickerView
            view.requiredText = requiredText
            view.isRequired = isRequired
            view.isValidationEnabled = isValidationEnabled
            view.showSelectedTextValue = showSelectedTextValue
            view.pickerDialog.restoreDialogState(dialogState)
            triggerState?.apply {
                view.triggerView?.onRestoreInstanceState(this)
            }
        }
    }

    fun interface SelectionChangedListener<SelectionType> {
        fun execute(newSelection: SelectionType?, oldSelection: SelectionType?)
    }

    fun interface ValidationCheck<SelectionType> {
        fun validate(currentSelection: SelectionType?): Boolean
    }

    fun interface TriggerAttachedCallback<SelectionType> {
        fun onAttached(triggerView: DialogTriggerView, pickerView: BaseDialogPickerView<SelectionType>)
    }

    fun interface DialogReadyListener<SelectionType> {
        fun execute(dialog: BasePickerDialogFragment<SelectionType>)
    }

    private class SavedState : BaseSavedState {
        var positiveButtonText: String = ""
        var negativeButtonText: String = ""
        var isDialogShown = false
        var requiredText: String? = null
        var isRequired = false
        var isValidationEnabled = false
        var showSelectedTextValue = false
        var dialogState: Bundle? = null
        var triggerState: Parcelable? = null

        constructor(superState: Parcelable?) : super(superState)

        private constructor(parcel: Parcel) : super(parcel) {
            parcel.apply {
                positiveButtonText = ParcelUtils.readString(this)!!
                negativeButtonText = ParcelUtils.readString(this)!!
                isDialogShown = ParcelUtils.readBoolean(this)!!
                isRequired = ParcelUtils.readBoolean(this)!!
                isValidationEnabled = ParcelUtils.readBoolean(this)!!
                showSelectedTextValue = ParcelUtils.readBoolean(this)!!
                requiredText = ParcelUtils.readString(this)
                dialogState = ParcelUtils.readParcelable(this, Bundle::class.java)
                triggerState = ParcelUtils.readParcelable(this, Parcelable::class.java)
            }
        }

        @Override
        override fun writeToParcel(parcel: Parcel, flags: Int) {
            super.writeToParcel(parcel, flags)
            parcel.apply {
                ParcelUtils.writeBoolean(this, isDialogShown)
                    .writeString(this, positiveButtonText)
                    .writeString(this, negativeButtonText)
                    .writeBoolean(this, isRequired)
                    .writeBoolean(this, isValidationEnabled)
                    .writeBoolean(this, showSelectedTextValue)
                    .writeString(this, requiredText)
                    .writeParcelable(this, dialogState)
                    .writeParcelable(this, triggerState)
            }
        }

        @Override
        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Creator<SavedState> {
            override fun createFromParcel(parcel: Parcel): SavedState {
                return SavedState(parcel)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }
        }
    }

    private enum class ButtonTypes(val value: Int) {
        POSITIVE(0),
        NEGATIVE(1)
    }
}